src/oracle/connection/stmt.rs

```diff
@@ -59,11 +58,29 @@ impl Statement {
                 "PREPARING STMT",
             )?;
 
-            // for create statements we need to run OCIStmtPrepare2 twice
-            // c.f. https://docs.oracle.com/database/121/LNOCI/oci17msc001.htm#LNOCI17165
-            // "To reexecute a DDL statement, you must prepare the statement again using OCIStmtPrepare2()."
-            if let Some(u) = mysql.to_string().find("CREATE") {
-                if u < 10 {
+            let stmt_type =
+                Self::get_statement_type(stmt, raw_connection.env.error_handle, &mysql)?;
+            is_returning = Self::is_returning(stmt, raw_connection.env.error_handle, &mysql)?;
+
+            // c.f. https://docs.oracle.com/database/121/LNOCI/oci04sql.htm#GUID-91AF021D-9FCD-4A4D-A647-2F2AB5B448B8__CIHEHCEJ
+            // c.f. https://stackoverflow.com/a/53390359/698496
+            match stmt_type as u32 {
```

Comment:
That match statement should be separated.
One for assigning `is_select` and one for calling `OCIStmtPrepare2` a second time.

Reply:
Hm, I don't see the benefit having this "monster" twice? Can you please elaborate.

Reply:
1. The "monster" will be smaller 
2. Just don't mix things that do not belong together...
3. At least the `is_select` part does not belong into the `unsafe` block

Something like this:
```rust
match stmt_type as u32 {
    ffi::OCI_STMT_SELECT => is_select = true
    ffi::OCI_STMT_UPDATE
   | ffi::OCI_STMT_DELETE
   | ffi::OCI_STMT_INSERT
   | ffi::OCI_STMT_BEGIN
   | ffi::OCI_STMT_DECLARE
   | 15
   | 16
   | 17
   | 21
   | ffi::OCI_ATTR_STMT_IS_RETURNING => is_select = false,
   _ => unreachable!("Statement type {} unknown", stmt_type),
}

if stmt_type as u32 ==  ffi::OCI_STMT_CREATE ||stmt_type as u32 ==  ffi::OCI_STMT_DROP ||stmt_type as u32 ==  ffi::OCI_STMT_ALTER {
     // for create statements we need to run OCIStmtPrepare2 twice
     // c.f. https://docs.oracle.com/database/121/LNOCI/oci17msc001.htm#LNOCI17165
     // "To reexecute a DDL statement, you must prepare the statement again using OCIStmtPrepare2()."

     let status = ffi::OCIStmtPrepare2(
          raw_connection.service_handle,
          &mut stmt,
          "PREPARING STMT 2"
      )?;
} 

```

Reply:
I think it belongs together:
* it prepares the statement itself and sets valid properties (either `is_select` will be set accordingly or `OCIStmtPrepare2` will be called depending on the statement type),
* if I leave out `ffi::OCI_STMT_CREATE | ffi::OCI_STMT_DROP | ffi::OCI_STMT_ALTER` from the match statement, we are going to meet the `unreachable` for each DDL statement.

Reply:
However I could refactor as:
```
let mut stmt = unsafe {
            let mut stmt: *mut ffi::OCIStmt = ptr::null_mut();
            let status = ffi::OCIStmtPrepare2(
                raw_connection.service_handle,
                &mut stmt,
                raw_connection.env.error_handle,
                mysql.as_ptr(),
                mysql.len() as u32,
                ptr::null(),
                0,
                ffi::OCI_NTV_SYNTAX,
                ffi::OCI_DEFAULT,
            );

            Self::check_error_sql(
                raw_connection.env.error_handle,
                status,
                &mysql,
                "PREPARING STMT",
            )?;

            stmt
        };

        let stmt_type = Self::get_statement_type(stmt, raw_connection.env.error_handle, &mysql)?;
        let is_returning = Self::is_returning(stmt, raw_connection.env.error_handle, &mysql)?;
        let is_select = stmt_type == ffi::OCI_STMT_SELECT;
        // c.f. https://docs.oracle.com/database/121/LNOCI/oci04sql.htm#GUID-91AF021D-9FCD-4A4D-A647-2F2AB5B448B8__CIHEHCEJ
        // c.f. https://stackoverflow.com/a/53390359/698496
        if stmt_type == ffi::OCI_STMT_CREATE
            || stmt_type == ffi::OCI_STMT_DROP
            || stmt_type == ffi::OCI_STMT_ALTER
        {
            unsafe {
                // for create statements we need to run OCIStmtPrepare2 twice
                // c.f. https://docs.oracle.com/database/121/LNOCI/oci17msc001.htm#LNOCI17165
                // "To reexecute a DDL statement, you must prepare the statement again using OCIStmtPrepare2()."

                let status = ffi::OCIStmtPrepare2(
                    raw_connection.service_handle,
                    &mut stmt,
                    raw_connection.env.error_handle,
                    mysql.as_ptr(),
                    mysql.len() as u32,
                    ptr::null(),
                    0,
                    ffi::OCI_NTV_SYNTAX,
                    ffi::OCI_DEFAULT,
                );

                Self::check_error_sql(
                    raw_connection.env.error_handle,
                    status,
                    &mysql,
                    "PREPARING STMT 2",
                )?;
            };
        }
```
Which I don't think is much better.